source/common/upstream/upstream_impl.cc

```diff
@@ -1125,6 +1125,11 @@ void StrictDnsClusterImpl::updateAllHosts(const HostVector& hosts_added,
                                           uint32_t current_priority) {
   PriorityStateManager priority_state_manager(*this, local_info_);
   // At this point we know that we are different so make a new host list and notify.
+  //
+  // TODO(dio): The uniqueness of a host address resolved in STRICT_DNS cluster per priority is not
+  // guaranteed. Need a clear agreement on the behavior here, whether it is allowable to have
+  // duplicated hosts inside a priority. And if we want to enforce this behavior, it should be done
```

Comment:
The other open question is whether we want to allow duplicated hosts *between* priorities. EDS does not allow this.

Reply:
I think duplicated hosts between priorities could land us in trouble, but I'm not sure if the scenario is plausible. Consider this:

host A: 50% of endpoints, duplicated between priority 0 and 1.
host B: 25% of endpoints, priority 0
host C: 25% of endpoints, priority 1

If a large portion of host A's endpoints are unhealthy and trigger some percentage of traffic to go to the priority 1 hosts, the overall health of priority 1 is affected as well and could trigger failover into an even lower priority. It seems unintuitive and unnecessary to allow duplicate hosts between priorities.

I'd like to hear an argument in the other direction if anyone has a use-case in mind.

Reply:
This has been brought up in other issues: https://github.com/envoyproxy/envoy/issues/4280#issuecomment-416800260a

The issue talks about EDS and would probably be supported by the indirection that's later suggested in the issue, but I'm not sure if we could do the same for STRICT_DNS.